src/open_options.rs

```diff
@@ -0,0 +1,139 @@
+    fn get_lints(&self) -> LintArray {
+        lint_array!(NONSENSICAL_OPEN_OPTIONS)
+    }
+}
+
+impl LateLintPass for NonSensicalOpenOptions {
+    fn check_expr(&mut self, cx: &LateContext, e: &Expr) {
+        if let ExprMethodCall(ref name, _, ref arguments) = e.node {
+            let (obj_ty, _) = walk_ptrs_ty_depth(cx.tcx.expr_ty(&arguments[0]));
+            if name.node.as_str() == "open" && match_type(cx, obj_ty, &OPEN_OPTIONS_PATH){
+                let mut options = Vec::new();
+                get_open_options(cx, &arguments[0], &mut options);
+                check_open_options(cx, &options, e.span);
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+enum Argument {
+    True,
+    False,
+    Unknown
+}
+
+#[derive(Debug)]
+enum OpenOption {
+    Write(Argument),
+    Read(Argument),
+    Truncate(Argument),
+    Create(Argument),
+    Append(Argument)
+}
+
+fn get_open_options(cx: &LateContext, argument: &Expr, options: &mut Vec<OpenOption>) {
+    if let ExprMethodCall(ref name, _, ref arguments) = argument.node {
+        let (obj_ty, _) = walk_ptrs_ty_depth(cx.tcx.expr_ty(&arguments[0]));
+        
+        // Only proceed if this is a call on some object of type std::fs::OpenOptions
+        if match_type(cx, obj_ty, &OPEN_OPTIONS_PATH) && arguments.len() >= 2 {
+            
+            let argument_option = match arguments[1].node {
+                ExprLit(ref span) => {
+                    if let Spanned {node: LitBool(lit), span: _} = **span {
+                        if lit {Argument::True} else {Argument::False}
+                    } else {
+                        return; // The function is called with a literal
+                                // which is not a boolean literal. This is theoretically
+                                // possible, but not very likely.
+                    }
+                },
+                _ => {
+                    Argument::Unknown
+                }
+            };
+            
+            match &*name.node.as_str() {
+                "create" => {
+                    options.push(OpenOption::Create(argument_option));
+                },
+                "append" => {
+                    options.push(OpenOption::Append(argument_option));
+                },
+                "truncate" => {
+                    options.push(OpenOption::Truncate(argument_option));
+                },
+                "read" => {
+                    options.push(OpenOption::Read(argument_option));
+                },
+                "write" => {
+                    options.push(OpenOption::Write(argument_option));
+                },
+                _ => {}
+            }
+            
+            get_open_options(cx, &arguments[0], options);
+        }
+    }
+}
+
+fn check_for_duplicates(cx: &LateContext, options: &[OpenOption], span: Span) {
+    // This code is almost duplicated (oh, the irony), but I haven't found a way to unify it.
```

Comment:
This would be easier if we have a plain `OpenOption enum` and use a `&[(OpenOption, Argument)]` slice instead.